docs/spec/namespace.md

```diff
@@ -0,0 +1,198 @@
+# Namespace
+The global namespace represents the full set of named repositories which can be
+referenced by Docker. Any scope within the namespace may refer to a subset of
+the named repos. A repository contains a set of content-addressable blobs and
+tags referencing those blobs. The namespace can be used to discover the
+location and certificate information of a repository based on DNS, HTTP
+requests and other means.
+
+A repository should always be referenced by its fully qualified name. If a
+client presents a shortened name to a user, that name should be fully expanded
+based on rules defined by the client before contacting a remote server. When
+mirroring repositories, the original name for a repository must be used.
+Changing a repository name is equivalent to copying or moving and will not be
+referenced by the original repository.
+
+## Terminology
+
+ - *Global Namespace* - The full set of referenceable names.
+ - *Name* - A fully qualified string containing both the domain and resource name
+ - *Repository* - A collection of objects under the same name within the
+namespace.
+ - *Namespace* - also *Namespace Scope* - A collection of repositories with a
+common name prefix and set of services including registry API, index, and trust
+context.
+ - *Short Name* - A name which does not contain a domain and requires
+expansion to a fully qualified name before resolving to a repository.
+
+## Format
+A name consists of two parts, a DNS host name plus a repository path. The host
+name follows the DNS rules without change. The total length of a name is 255
+characters however there is no specific limitation on DNS or path components.
+
+### Name Grammar
+```
+<name> ::= <hostname>"/"<path>
+<hostname> ::= <host-part>*["."<host-part>][":"<port>]
+<path> ::= <path-part>*["/"<path-part>]
+<host-part> ::= <regexp "[a-z]([-]?[a-z0-9])*">
```

Comment:
Ok, I would allow starting with a number, but requiring a letter somewhere after that.   We should also allow multiple hyphens especially for punycode.

```diff
- <host-part> ::= <regexp "[a-z]([-]?[a-z0-9])*">
+ <host-part> ::= <regexp "([0-9][-0-9]*)?[a-z](-*[a-z0-9])*">
# Number, followed by optional non-letters, followed by required letter, followed by anything, then a non-hyphen character to end

valid:
9-a
xn--n3h
lib

invalid:
9
99-99
a0- (still invalid)
```